### PR TITLE
fix: [CDS-68390]: add logic to refactor Kubectl client

### PR DESCRIPTION
**Pull request by achyuth-harness**
## Describe your changes
Adding KubectlFactory that will be used for creating Kubectl and Openshift clients
This factory will preserve a version of the client that later can be used for decisions on which commands should be called based on client versions
Introducing OcClient as well which uses commands implemented in Kubectl but in case we want to override any of those in the future we can easily do it
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/47730)

### CODE REVIEW

**Comment by alebuh** [comment]
trigger uts

**Comment by alebuh** [comment]
trigger ut4

File: 960-api-services/src/main/java/io/harness/k8s/kubectl/KubectlFactory.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  private KubectlFactory() {
+    throw new IllegalStateException("Utility class");
+  }
+
+  public static Kubectl getClient(String kubectlPath, String configPath, String workingDir) {
+    Kubectl kubectl = Kubectl.client(kubectlPath, configPath);
+    VersionCommand versionCommand = kubectl.version();
+    String version;
+    try {
+      version = versionCommand.execute(workingDir, null, null, false, Collections.emptyMap()).outputUTF8();
```
Comment (abosii-harness) [inline 960-api-services/src/main/java/io/harness/k8s/kubectl/KubectlFactory.java:33]: We may need to add client=true to version command

File: 960-api-services/src/main/java/io/harness/k8s/kubectl/KubectlFactory.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+      log.debug(format("Kubectl client with version %s has been created", version));
+    } catch (Exception ex) {
+      log.error("Failed to get kubernetes version, defaulting to version: " + DEFAULT_KUBECTL_VERSION, ex);
```
Comment (abosii-harness) [inline 960-api-services/src/main/java/io/harness/k8s/kubectl/KubectlFactory.java:36]: This should be infor or at least warn

**Review by abosii-harness — CHANGES_REQUESTED** [review]

**Review by abosii-harness — CHANGES_REQUESTED** [review]
// ignore

**Comment by alebuh** [comment]
trigger uts

File: 960-api-services/src/main/java/io/harness/k8s/kubectl/KubectlFactory.java (hunk 2 above)
Comment (alebuh) [reply 960-api-services/src/main/java/io/harness/k8s/kubectl/KubectlFactory.java]: done

File: 960-api-services/src/main/java/io/harness/k8s/kubectl/KubectlFactory.java (hunk 1 above)
Comment (alebuh) [reply 960-api-services/src/main/java/io/harness/k8s/kubectl/KubectlFactory.java]: done

**Comment by alebuh** [comment]
trigger uts

**Comment by alebuh** [comment]
trigger uts

**Comment by alebuh** [comment]
trigger uts

**Review by markoharness — APPROVED** [review]
> Reviewed 1 of 4 files at r4, 6 of 6 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47730)* status: 8 file(s) to review. 2. unresolved discussions.

**Comment by alebuh** [comment]
trigger review

**Pull request merged**